lambdas/pkgpush/index.py

```
@@ -0,0 +1,309 @@
+PKG_FROM_FOLDER_MAX_FILES = PROMOTE_PKG_MAX_FILES
+
+
+user_boto_session = None
+
+
+class S3ClientProvider(quilt3.data_transfer.S3ClientProvider):
+    def get_boto_session(self):
+        return user_boto_session
+
+
+quilt3.data_transfer.S3ClientProvider = S3ClientProvider
+
+
+@contextlib.contextmanager
+def setup_user_boto_session(**kwargs):
+    global user_boto_session
+    user_boto_session = boto3.session.Session(**kwargs)
+    try:
+        yield user_boto_session
+    finally:
+        user_boto_session = None
+
+
+class ApiException(Exception):
+    pass
+
+
+def get_registry(registry_url):
+    package_registry = None
+    try:
+        package_registry = get_package_registry(registry_url)
+    except quilt3.util.URLParseError:
+        pass
+    else:
+        if not isinstance(package_registry, S3PackageRegistryV1):
+            package_registry = None
+    if package_registry is None:
+        raise ApiException(HTTPStatus.BAD_REQUEST, f"{registry_url} is not a valid S3 package registry.")
+    return package_registry
+
+
+def _get_successor_params(registry, successor):
+    workflow_validator = registry.get_workflow_validator()
+    successors = workflow_validator.config.get('successors') or {}
+    for successor_url, successor_params in successors.items():
+        if get_registry(successor_url) == successor:
+            return successor_params
+    raise ApiException(HTTPStatus.BAD_REQUEST, f"{successor.base} is not configured as successor.")
+
+
+def _push_pkg_to_successor(data, *, get_src, get_dst, get_name, get_pkg, pkg_max_size, pkg_max_files):
+    dst_registry = get_registry(get_dst(data))
+    src_registry = get_registry(get_src(data))
+    copy_data = _get_successor_params(src_registry, dst_registry).get('copy_data', True)
+
+    try:
+        pkg = get_pkg(src_registry, data)
+        if copy_data:
+            total_size = 0
+            total_files = 0
+            for lk, e in pkg.walk():
```

It would be much cleaner to return total_size, and total_files as a tuple from a utility function that takes the package as an argument.

---

I think it doesn't make sense because we want to fail ASAP and in future we'd want to rework it to fail without loading the whole manifest in the memory.

---

I can live with it but wish to point out that you can still fail early even if the iteration happens in a separate function

---

Well, currently a function where iteration happens is `Package.set_dir()`.